### PR TITLE
[Dynamic Instrumentation] DEBUG-4089 Support file-based debugger probes

### PR DESCRIPTION
**Pull request by dudikeleti**
## Summary of changes
Added support to load log/metric/span/span‑decoration probes from a JSON file (DebuggerSettings.ProbeFile)

## Reason for change
Dynamic Instrumentation only worked with RCM‑delivered probes. There was no support for loading probes from a local file, and the internal configuration state wasn’t designed for combining file and RCM probes or handling file‑only 

## Implementation details

- Added support to load log/metric/span/span‑decoration probes from a JSON file (DebuggerSettings.ProbeFile) during DynamicInstrumentation initialization.
- Ensured Dynamic Instrumentation can operate in three modes: RCM‑only, file‑only, and combined file+RCM.
- Implemented source‑level merge (MergeProbes) where RCM wins on ID conflicts with file probes, and configuration‑level union (MergeConfigurations).
- Updated initialization to start background processing and status/snapshot upload when either RCM is available or file probes exist.

## Test coverage
Updated DynamicInstrumentationTests to cover file probe loading (multiple probe types, invalid/missing/empty files, no file configured, partially valid file entries, and duplicate IDs within a file) and merge semantics between file and RCM probes.

### CODE REVIEW

File: tracer/src/Datadog.Trace/Debugger/Configurations/ConfigurationUpdater.cs [hunk 1]
```diff
@@ -20,26 +20,85 @@ internal sealed class ConfigurationUpdater
         private readonly string? _env;
         private readonly string? _version;
         private readonly int _maxProbesPerType;
+        private readonly HashSet<string> _removedRcmProbeIds = new(StringComparer.Ordinal);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ConfigurationUpdater.cs]: nit: ordinal is the default, and it's better to not specify it in these cases, as there are often internal fast-paths in the types 🙂 

```suggestion
        private readonly HashSet<string> _removedRcmProbeIds = new();
```

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        {
+            return higherPriority;
+        }
+
+        if (higherPriority.Length == 0)
+        {
+            return lowerPriority;
+        }
+
+        var mergedProbes = new Dictionary<string, T>(StringComparer.Ordinal);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs]: nit: Use default comparer where possible
```suggestion
        var mergedProbes = new Dictionary<string, T>();
```

File: tracer/src/Datadog.Trace/Debugger/DynamicInstrumentation.cs [hunk 3]
```diff
@@ -556,21 +590,21 @@ private ApplyDetails[] AcceptAddedConfiguration(List<RemoteConfiguration>? confi
...
             };
 
             try
             {
-                var updateResults = _configurationUpdater.AcceptAdded(probeConfiguration);
+                var updateResults = _configurationUpdater.AcceptAdded(rcmUpdate);
                 foreach (var updateResult in updateResults)
                 {
-                    var config = configs.FirstOrDefault(c => c.Path.Id == updateResult.Id);
+                    var config = configs.FirstOrDefault(c => ProbeConfigurationUtils.GetProbeIdFromPath(c.Path) == updateResult.Id);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/DynamicInstrumentation.cs]: `GetProbeIdFromPath` is very allocatey, as it creates a substring. Given you're calling this for every config, in every result, that's a lot of extra strings when you just are looking for the config.

I'd suggest rewriting `GetProbeIdFromPath()` to either return a `ReadOnlySpan<char>` which you can use here for the comparison without allocating, or instead create a `IsProbeId(path, id)` method that does something similar

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    private static T[] DeduplicateProbes<T>(List<T> probes)
+        where T : ProbeDefinition
+    {
+        if (probes.Count == 0)
+        {
+            return [];
+        }
+
+        var uniqueProbes = new List<T>(probes.Count);
+        var seenIds = new HashSet<string>(StringComparer.Ordinal);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]:
```suggestion
        var seenIds = new HashSet<string>();
```

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                fileContent = await reader.ReadToEndAsync().ConfigureAwait(false);
+            }
+
+            if (string.IsNullOrWhiteSpace(fileContent))
+            {
+                Log.Debug("Probe file is empty: {ProbeFile}", probeFile);
+                return null;
+            }
+
+            var jArray = JsonHelper.ParseJArray(fileContent);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]: This reads everything into a string, and then parses that string: it's better to parse directly rom the file stream, as it avoids the intermediate `string` allocation (which is just thrown away)

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            var logs = new List<LogProbe>();
+            var metrics = new List<MetricProbe>();
+            var spans = new List<SpanProbe>();
+            var spanDecorations = new List<SpanDecorationProbe>();
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]: nit: may be better to initialize these to null, and then do logs ??= new(); before the `Add()` calls, to reduce some allocation

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+                    continue;
+                }
+
+                var type = typeToken.ToString();
+                try
+                {
+                    switch (type)
+                    {
+                        case "LOG_PROBE":
+                            var logProbe = jObject.ToObject<LogProbe>();
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]: Is there a `JsonHelper` version of this we can use instead, so that we use the array pool? 🤔

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            }
+
+            var totalProbes = logs.Count + metrics.Count + spans.Count + spanDecorations.Count;
+            if (totalProbes == 0)
+            {
+                Log.Warning("No valid probes found in file: {ProbeFile}", probeFile);
+                return null;
+            }
+
+            var uniqueLogs = DeduplicateProbes(logs);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]: Why not de-duplicate when you're adding, instead of creating a whole new list? 🤔 I think the behaviour would be the same.

Also, if you don't actually expect that many items in the file, then it may not be worth the overhead of creating a hashset, and using a simple `Contains()` on the `List<>` may be preferable

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        if (removedProbeIds.Count == 0 && !removeServiceConfiguration)
+        {
+            return configuration;
+        }
+
+        return new ProbeConfiguration
+        {
+            ServiceConfiguration = removeServiceConfiguration ? null : configuration.ServiceConfiguration,
+            LogProbes = configuration.LogProbes.Where(probe => !removedProbeIds.Contains(probe.Id)).ToArray(),
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs]: This feels very allocate-y, lots of Linq and `ToArray()` calls in many places - is that necessary? 🤔 Wondering if we can either refactor to just mutate lists instead of creating many instances of arrays etc

File: tracer/src/Datadog.Trace/Debugger/Configurations/ConfigurationUpdater.cs
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ConfigurationUpdater.cs:1]: I feel like this file should be heavily unit-tested, independent of the "integration" tests in DynamicInstrumentation, as most of the behaviour should be "given inputs, expect these outputs"

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs:1]: I feel like this file should be heavily unit-tested, independent of the "integration" tests in DynamicInstrumentation, so that you can test all the edge cases

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs:1]: The methods in this class are another example of things that should be independently unit tested, so that we cover all the edge cases

**Review by andrewlock — APPROVED** [review]
Approving for the tracer side - I glanced over some of the debugger stuff, mostly just to point out a few opportunities for perf improvements. One general comment is that I think most of the "helper" classes should all have their own dedicated unit tests for their behaviour, so we can cover all the edge cases, without going through the "full" dynamic instrumentation initialization code

File: tracer/src/Datadog.Trace/Debugger/Configurations/ConfigurationUpdater.cs (hunk 1 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ConfigurationUpdater.cs]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs (hunk 2 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/DynamicInstrumentation.cs (hunk 3 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/DynamicInstrumentation.cs]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs (hunk 4 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs (hunk 5 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs (hunk 7 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs (hunk 8 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs (hunk 9 above)
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationFileLoader.cs:1]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ProbeConfigurationUtils.cs:1]: dca07cf75d8aa42be99374d4028d5593240b3216

File: tracer/src/Datadog.Trace/Debugger/Configurations/ConfigurationUpdater.cs
Comment (dudikeleti) [reply tracer/src/Datadog.Trace/Debugger/Configurations/ConfigurationUpdater.cs:1]: dca07cf75d8aa42be99374d4028d5593240b3216

**Review by tylfin — APPROVED** [review]
